File: src/main/java/no/nav/bidrag/beregn/saertilskudd/rest/mapper/SamvaersfradragCoreMapper.java
package no.nav.bidrag.beregn.saertilskudd.rest.mapper;

import java.util.ArrayList;
import no.nav.bidrag.beregn.saertilskudd.rest.consumer.SjablonListe;
import no.nav.bidrag.beregn.saertilskudd.rest.dto.http.BeregnTotalSaertilskuddGrunnlag;
import no.nav.bidrag.beregn.saertilskudd.rest.dto.http.Grunnlag;
import no.nav.bidrag.beregn.saertilskudd.rest.dto.http.GrunnlagType;
import no.nav.bidrag.beregn.saertilskudd.rest.dto.http.Samvaersklasse;
import no.nav.bidrag.beregn.samvaersfradrag.dto.BeregnSamvaersfradragGrunnlagCore;
import no.nav.bidrag.beregn.samvaersfradrag.dto.SamvaersklassePeriodeCore;

public class SamvaersfradragCoreMapper extends CoreMapper {

  public BeregnSamvaersfradragGrunnlagCore mapSamvaersfradragGrunnlagTilCore(BeregnTotalSaertilskuddGrunnlag beregnTotalSaertilskuddGrunnlag,
      SjablonListe sjablonListe) {

    var samvaersklassePeriodeCoreListe = new ArrayList<SamvaersklassePeriodeCore>();

    // Løper gjennom alle grunnlagene og identifiserer de som skal mappes til samværsfradrag core
    for (Grunnlag grunnlag : beregnTotalSaertilskuddGrunnlag.getGrunnlagListe()) {
      if (GrunnlagType.SAMVAERSKLASSE.equals(grunnlag.getType())) {
        Samvaersklasse samvaersklasse = grunnlagTilObjekt(grunnlag, Samvaersklasse.class);
        samvaersklassePeriodeCoreListe.add(samvaersklasse.tilCore(grunnlag.getReferanse()));
      }
    }

    // Henter aktuelle sjabloner
    var sjablonPeriodeCoreListe = new ArrayList<>(mapSjablonSamvaersfradrag(sjablonListe.getSjablonSamvaersfradragResponse(),
        beregnTotalSaertilskuddGrunnlag));

    return new BeregnSamvaersfradragGrunnlagCore(beregnTotalSaertilskuddGrunnlag.getBeregnDatoFra(),
        beregnTotalSaertilskuddGrunnlag.getBeregnDatoTil(), samvaersklassePeriodeCoreListe, sjablonPeriodeCoreListe);
  }
}

File: src/main/kotlin/no/nav/bidrag/beregn/saertilskudd/rest/mapper/SamvaersfradragCoreMapper.kt
package no.nav.bidrag.beregn.saertilskudd.rest.mapper

import no.nav.bidrag.beregn.saertilskudd.rest.consumer.SjablonListe
import no.nav.bidrag.beregn.saertilskudd.rest.dto.http.BeregnTotalSaertilskuddGrunnlag
import no.nav.bidrag.beregn.saertilskudd.rest.dto.http.GrunnlagType
import no.nav.bidrag.beregn.saertilskudd.rest.dto.http.Samvaersklasse
import no.nav.bidrag.beregn.samvaersfradrag.dto.BeregnSamvaersfradragGrunnlagCore
import no.nav.bidrag.beregn.samvaersfradrag.dto.SamvaersklassePeriodeCore

object SamvaersfradragCoreMapper : CoreMapper() {
    fun mapSamvaersfradragGrunnlagTilCore(
        beregnTotalSaertilskuddGrunnlag: BeregnTotalSaertilskuddGrunnlag,
        sjablonListe: SjablonListe
    ): BeregnSamvaersfradragGrunnlagCore {
        val samvaersklassePeriodeCoreListe = ArrayList<SamvaersklassePeriodeCore>()

        // Løper gjennom alle grunnlagene og identifiserer de som skal mappes til samværsfradrag core
        for (grunnlag in beregnTotalSaertilskuddGrunnlag.grunnlagListe!!) {
            if (GrunnlagType.SAMVAERSKLASSE == grunnlag.type) {
                val samvaersklasse = grunnlagTilObjekt(grunnlag, Samvaersklasse::class.java)
                samvaersklassePeriodeCoreListe.add(samvaersklasse.tilCore(grunnlag.referanse!!))
            }
        }

        // Henter aktuelle sjabloner
        val sjablonPeriodeCoreListe = ArrayList(
            mapSjablonSamvaersfradrag(
                sjablonListe.sjablonSamvaersfradragResponse,
                beregnTotalSaertilskuddGrunnlag
            )
        )
        return BeregnSamvaersfradragGrunnlagCore(
            beregnTotalSaertilskuddGrunnlag.beregnDatoFra!!,
            beregnTotalSaertilskuddGrunnlag.beregnDatoTil!!, samvaersklassePeriodeCoreListe, sjablonPeriodeCoreListe
        )
    }
}
